metron-interface/metron-rest/src/main/java/org/apache/metron/rest/service/impl/AlertServiceImpl.java

```diff
@@ -0,0 +1,57 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.metron.rest.service.impl;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import java.util.Map;
+import org.apache.metron.common.utils.JSONUtils;
+import org.apache.metron.rest.MetronRestConstants;
+import org.apache.metron.rest.RestException;
+import org.apache.metron.rest.service.AlertService;
+import org.apache.metron.rest.service.KafkaService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.env.Environment;
+import org.springframework.stereotype.Service;
+
+/**
+ * The default service layer implementation of {@link AlertService}.
+ *
+ * @see AlertService
+ */
+@Service
+public class AlertServiceImpl implements AlertService {
+
+  private Environment environment;
+  private final KafkaService kafkaService;
+
+  @Autowired
+  public AlertServiceImpl(final KafkaService kafkaService,
+                          final Environment environment) {
+    this.kafkaService = kafkaService;
+    this.environment = environment;
+  }
+
```

Comment:
Is this doing a single alert at a time?
Does it actually pass all the alert data from the client to the endpoint?

Reply:
I can adjust it to accept a list of alerts.  I agree that would be better.

Yes the endpoint expects all alert data to be passed in.  I'm trying to avoid an extra ES call (alert data will travel across the network either way).  

But now that you bring it up I'm not sure.  Do we want to make it flexible so that an alert can be annotated/modified before being escalated?  Would fetching an alert from ES be unnecessary since it's assumed the client will already have a copy?  Would we want to ensure that the most recent version in ES is always escalated?  Should we expose both options?

Reply:
What I was thinking about is the assumption that the client will have the 'full alert'.
If the client requests alerts based on queries, and specifies the fields for return as a subset of the complete fields, then you are making the assumption that the subset is what will be required for escalation.  These are different use cases.  It is more likely that what will be required is:

- Query from any point to service returned sum subset of fields for alert A.
- User decides to escalate, and has some option to add new fields/data/annotations to alerts  ( best would we to have some 'additional field list + description' per escalation target that we can load into the ui ) to alert.
- User may add same fields/annotations to multiple alerts ( escalate all in query )
- Escalation takes annotations and alert id's, merges the annotations with the full alert (or from a query that returns the fields required by the target?) and sends them

Reply:
I was coming at it from a transactional perspective and trying to avoid a race condition (alert changes after a client requests it and before it gets escalated).  I think a lot of what you describe can already be done by chaining rest calls together.

Reply:
Can you give an example?  I don't see how.  

escalate(List<?> alertId, String patchFields)?

When we get the alerts.... is there a version - so that we will just work with the version we have and even if v+1 goes in it won't matter?

Reply:
My understanding is that this is for manual escalation, and should just expose everything in the alert (or meta-alert) that is being manually escalated. 

The user would essentially add comments and make modifications before escalating, which would be persisted 

The alert which is pushed onto the queue should be the current alert state, i.e. with all those modifications. This was done elsewhere (sorry, can't remember the ticket number). So the alert data being pushed should come off the cluster side, and it should have all the content already sent from the client by a separate process. 

I think that should achieve pretty much what you're talking about @ottobackwards.

Reply:
@simonellistonball I don't think that is what this api does.  We need to clarify.
There is also a difference between the rest api and the ui workflow.  They both have to make sense on their own.

Do we need a discussion thread?  or a document?

Reply:
Documentation issues aside, are we concerned about a potential race condition?  It sounds like we are.  In that case we would need to retrieve the alerts from ES in the REST service before they are escalated.  Is my understanding correct?

Reply:
I wouldn't use the word concerned.  I am interested in the use cases and api design.

Reply:
> Yes the endpoint expects all alert data to be passed in. I'm trying to avoid an extra ES call (alert data will travel across the network either way).

I think @merrimanr 's implementation makes the most sense right now.  We assume the client has all of the necessary details about the alert that they want to escalate.  Adding extra logic to pull the freshest version from the index (ES) does not make sense to me.

As a SOC analyst using the Alerts UI, if I escalate an alert, I want to know that what I am seeing on my screen is what I actually escalated.  I escalated the alert based on the information that I had in front of me; it is a point-in-time snapshot.

If the alert changes between the time that I loaded the alert on my screen and when I clicked escalate, then that is a condition that can only be correctly recovered from in the UI, not in the REST API.  

In addition, the result of escalation here is that we are pushing data into a Kafka topic.  The only reasonable expectation is that this data is a point-in-time snapshot.  We don't need to jump through hoops to try and sync with any changes in the index.

If the alerts data needs refreshed, then the consumer of the 'escalation' Kafka topic needs to do that.  Whether the data needs refreshed and how fresh that needs to be are decisions that only the consumer of that topic can make.